In [1]:
# Install the numpy library if it's not already installed
pip install numpy

In [2]:
# Import the numpy library and print its version
import numpy as np
print(np.__version__)

2.0.2


In [3]:
import time
import numpy as np

# Define the size of the lists/arrays for performance comparison
size = 10**6

# --- Python List comparison ---
# Record start time for Python list operations
start_time = time.time()
# Create two Python lists
py_list1 = list(range(size))
py_list2 = list(range(size))
# Perform element-wise addition using a list comprehension
py_result = [py_list1[i] + py_list2[i] for i in range(size)]
# Record end time and calculate duration for Python list
end_time = time.time()
python_list_time = end_time - start_time
print(f"Time taken for Python List operation: {python_list_time:.6f} seconds")

# --- NumPy Array comparison ---
# Record start time for NumPy array operations
start_time = time.time()
# Create two NumPy arrays
numpy_array1 = np.arange(size)
numpy_array2 = np.arange(size)
# Perform element-wise addition using NumPy's vectorized operation
numpy_result = numpy_array1 + numpy_array2
# Record end time and calculate duration for NumPy array
end_time = time.time()
numpy_array_time = end_time - start_time
print(f"Time taken for NumPy Array operation: {numpy_array_time:.6f} seconds")

# Compare the performance of NumPy versus Python lists
print(f"\nNumPy is approximately {python_list_time / numpy_array_time:.2f} times faster than Python list for this operation.")

Time taken for Python List operation: 0.149994 seconds
Time taken for NumPy Array operation: 0.008269 seconds

NumPy is approximately 18.14 times faster than Python list for this operation.


In [ ]:
# This cell is empty.

In [6]:
import sys

# Calculate memory usage for Python list
# sys.getsizeof() gets the size of the list object itself, then sum of sizes of each element
memory_py_list = sys.getsizeof(py_list1) + sum(sys.getsizeof(item) for item in py_list1)
print(f"Memory taken by Python List: {memory_py_list / (1024**2):.4f} MB")

# Calculate memory usage for NumPy array
# .nbytes gives the total bytes consumed by the array elements
memory_numpy_array = numpy_array1.nbytes
print(f"Memory taken by NumPy Array: {memory_numpy_array / (1024**2):.4f} MB")

# Compare memory usage between Python list and NumPy array
print(f"\nPython List uses approximately {memory_py_list / memory_numpy_array:.2f} times more memory than NumPy array for this data.")

Memory taken by Python List: 34.3323 MB
Memory taken by NumPy Array: 7.6294 MB

Python List uses approximately 4.50 times more memory than NumPy array for this data.


## Summary of Time and Memory Comparison

Based on the operations performed:

### Time Performance
*   **Python List**: `0.1500` seconds
*   **NumPy Array**: `0.0083` seconds

NumPy Arrays were approximately **`18.14` times faster** than Python lists for the element-wise addition operation on `10^6` elements.

### Memory Usage
*   **Python List**: `34.3323` MB
*   **NumPy Array**: `7.6294` MB

Python Lists used approximately **`4.50` times more memory** than NumPy arrays for storing `10^6` integer elements. This is because Python lists store references to objects, while NumPy arrays store contiguous blocks of homogeneous data.

In [91]:
# Define a Python list and check its type
list = [89,1,2,3,4,5]
type(list)

list

In [92]:
# Convert the Python list to a NumPy array
arr1=np.array(list)

In [93]:
# Print the NumPy array
print(arr1)

[89  1  2  3  4  5]


In [94]:
# Check the type of the NumPy array
type(arr1)

numpy.ndarray

In [95]:
# Get the number of dimensions of the array
arr1.ndim

1

In [96]:
# Get the data type of the array elements
arr1.dtype

dtype('int64')

In [97]:
# Get the name of the data type of the array elements
arr1.dtype.name

'int64'

In [98]:
# Get the total number of elements in the array
arr1.size

6

In [99]:
# Get the total number of elements in the array (duplicate cell)
arr1.size

6

In [100]:
# Get the shape of the array (dimensions of the array)
arr1.shape

(6,)

In [101]:
# Get the total bytes consumed by the array elements
arr1.nbytes

48

In [103]:
# Define a Python tuple
tuple=(89,1,2,3,5)

In [104]:
# Check the type of the variable 'tuple'
type(tuple)

tuple

In [105]:
# Define another Python tuple with a single element to demonstrate its structure
tuple2=(1,)

In [106]:
# Print the second tuple
tuple2

(1,)

In [107]:
# Check the type of the second tuple
type(tuple2)

tuple

In [108]:
# Convert the first tuple to a NumPy array
arr2=np.array(tuple)

In [109]:
# Print the NumPy array created from the tuple
print(arr2)

[89  1  2  3  5]


In [110]:
# Get the number of dimensions of arr2
arr2.ndim

1

In [113]:
# Define a nested Python list, representing a 2D structure
list=[[89,2],[3,4]]

In [114]:
# Convert the nested list to a NumPy matrix (2D array)
matrix=np.array(list)

In [115]:
# Print the NumPy matrix
print(matrix)

[[89  2]
 [ 3  4]]


In [116]:
# Get the number of dimensions of the matrix
matrix.ndim

2

In [130]:
# Get the shape of the matrix (rows, columns)
matrix.shape

(2, 2)

In [119]:
# Get the total number of elements in the matrix
matrix.size

4

In [131]:
# Get the total bytes consumed by the matrix elements
matrix.nbytes

32

In [132]:
# Flatten the matrix into a 1D array (returns a copy)
matrix.flatten()

array([89,  2,  3,  4])

In [133]:
# Flatten the matrix into a 1D array (returns a view if possible, otherwise a copy)
matrix.ravel()

array([89,  2,  3,  4])

In [134]:
# Create a 1D NumPy array directly
arr4=np.array([89,2,3,4])

In [127]:
# Print the 1D NumPy array
print(arr4)

[89  2  3  4]


In [129]:
# Create a 2D NumPy array directly
arr5=np.array([[89,2],[3,4]])

In [45]:
# Print the 2D NumPy array (Note: The output seems to be from an older execution or a different variable. This cell's content is `print(arr5)` and `arr5` should be `[[89, 2], [3, 4]]` based on the previous edit.)
print(arr5)

[[1 2]
 [3 4]]


In [136]:
# Create a NumPy matrix (2D array) using `np.matrix`
arr6=np.matrix([[89,2],[3,4],[5,6]])

In [137]:
# Print the NumPy matrix arr6
print(arr6)

[[89  2]
 [ 3  4]
 [ 5  6]]


In [138]:
import numpy as np

# Create a 1D array with numbers from 1 to 100 using np.arange
array_1_to_100 = np.arange(1, 101)

# Print the array
print(array_1_to_100)

[  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
  91  92  93  94  95  96  97  98  99 100]


In [50]:
# Create a 1D array with numbers from 1 to 99 (exclusive of 100)
arr7=np.arange(1,100)

In [51]:
# Print the array arr7
print(arr7)

[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48
 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72
 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96
 97 98 99]


In [56]:
# Create an array with even numbers from 0 to 100 (exclusive of 101), with a step of 2
arr8=np.arange(0,101,2)

In [57]:
# Print the array arr8
print(arr8)

[  0   2   4   6   8  10  12  14  16  18  20  22  24  26  28  30  32  34
  36  38  40  42  44  46  48  50  52  54  56  58  60  62  64  66  68  70
  72  74  76  78  80  82  84  86  88  90  92  94  96  98 100]


In [58]:
# Create a 1D array with numbers from 0 to 9
matrix2=np.arange(0,10)

In [59]:
# Print the 1D array matrix2
print(matrix2)

[0 1 2 3 4 5 6 7 8 9]


In [60]:
# Reshape the 1D array (0-9) into a 2x5 matrix
matrix2=np.arange(0,10).reshape(2,5)

In [61]:
# Print the reshaped 2x5 matrix
print(matrix2)

[[0 1 2 3 4]
 [5 6 7 8 9]]


In [62]:
# Create a 3x3 array filled with zeros (default dtype is float)
a=np.zeros((3,3))

In [63]:
# Print the array 'a'
print(a)

[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]


In [64]:
# Get the name of the data type of elements in array 'a'
a.dtype.name

'float64'

In [66]:
# Create a 3x3 array filled with zeros, explicitly setting the data type to integer
a=np.zeros((3,3),dtype='int')

In [67]:
# Print the integer array 'a'
print(a)

[[0 0 0]
 [0 0 0]
 [0 0 0]]


In [68]:
# Create a 3x3 array filled with ones, explicitly setting the data type to integer
b=np.ones((3,3),dtype='int')

In [69]:
# Print the array 'b' (filled with ones)
print(b)

[[1 1 1]
 [1 1 1]
 [1 1 1]]


In [73]:
# Create a 10x10 array filled with a specific value (5)
c=np.full((10,10),5)

In [74]:
# Print the array 'c' (filled with fives)
print(c)

[[5 5 5 5 5 5 5 5 5 5]
 [5 5 5 5 5 5 5 5 5 5]
 [5 5 5 5 5 5 5 5 5 5]
 [5 5 5 5 5 5 5 5 5 5]
 [5 5 5 5 5 5 5 5 5 5]
 [5 5 5 5 5 5 5 5 5 5]
 [5 5 5 5 5 5 5 5 5 5]
 [5 5 5 5 5 5 5 5 5 5]
 [5 5 5 5 5 5 5 5 5 5]
 [5 5 5 5 5 5 5 5 5 5]]


In [75]:
# Create a 3x3 identity matrix
e=np.eye(3)

In [76]:
# Print the identity matrix 'e'
print(e)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [77]:
# Create an array with 5 evenly spaced numbers between 0 and 1 (inclusive)
f=np.linspace(0,1,5)

In [78]:
# Print the array 'f'
print(f)

[0.   0.25 0.5  0.75 1.  ]


In [80]:
# Import the random module as 'random' (Note: NumPy also has its own random module)
import random as random

In [82]:
# Create a 3x3 array with random floating-point numbers between 0.0 and 1.0
g=np.random.random((3,3))

In [83]:
# Print the array 'g' containing random numbers
print(g)

[[0.40569468 0.74919459 0.06846254]
 [0.89274149 0.65299153 0.4861576 ]
 [0.27546159 0.1236944  0.01220258]]


In [86]:
# Create a 1D array with numbers from 1 to 9
arr11=np.arange(1,10)

In [88]:
# Print the array arr11
print(arr11)

[1 2 3 4 5 6 7 8 9]


In [89]:
# Find and display the maximum value in array arr11
arr11.max()

np.int64(9)

In [ ]:
# This cell is currently empty.